snapcraft/internal/meta/_desktop.py

```diff
@@ -57,14 +57,14 @@ def parse_and_reformat(self, extracted_metadata: _metadata.ExtractedMetadata):
             exec_value = "{}.{} %U".format(self._snap_name, self._name)
         self._parser[section]["Exec"] = exec_value
         if "Icon" in self._parser[section]:
+            icon = self._parser[section]["Icon"]
+
             # Extracted metadata (e.g. from the AppStream) can override the
             # icon location.
             if extracted_metadata:
-                metadata_dict = extracted_metadata.to_dict()
-                if "icon" in metadata_dict:
-                    icon = "/" + str(metadata_dict["icon"])
-            else:
-                icon = self._parser[section]["Icon"]
+                metadata_icon = extracted_metadata.get_icon()
+                if metadata_icon is not None:
+                    icon = "/" + metadata_icon
```

Comment:
I wonder why we add this slash here to strip it a line below. :man_shrugging: 
Not that you did this, but maybe worth a fix :smile:

Reply:
if metadata_icon even had a leading /, by adding one, the logic below makes no sense.

Reply:
Indeed it doesn't. Refactored existing code to remove unnecessary maneuvering with initial slash.

Reply:
Oh, there is a situation where the original code makes sense: stock icons in classic confinement. In this case, the base name of the icon is specified instead of the full path and the system will search on the standard system theme locations.